component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -161,58 +160,191 @@ protected void initiateAuthenticationRequest(HttpServletRequest request,
+            if (StringUtils.isEmpty(usecase)) {
+                Map<Integer, StepConfig> stepConfigMap = context.getSequenceConfig().getStepMap();
+                Map<ClaimMapping, String> userAttributes = new HashMap<>();
+                String federatedEmailAttributeKey = null;
+                String username = null;
+                boolean isLocalUser = false;
+                // Iterate through the steps to identify from which step the user email address need to extracted
+                for (StepConfig stepConfig : stepConfigMap.values()) {
+                    authenticatedUser = stepConfig.getAuthenticatedUser();
+                    if (authenticatedUser != null && stepConfig.isSubjectAttributeStep()) {
+                        username = authenticatedUser.getUserName();
+                        if (stepConfig.getAuthenticatedAutenticator().getApplicationAuthenticator() instanceof
+                                LocalApplicationAuthenticator) {
+                            isLocalUser = true;
+                            break;
+                        }
+                        userAttributes = authenticatedUser.getUserAttributes();
+                        federatedEmailAttributeKey = getFederatedEmailAttributeKey(context,
+                                stepConfig.getAuthenticatedAutenticator().getName());
+                        break;
+                    }
                 }
-                processEmailOTPMandatory(context, request, response, isUserExistence, username, queryParams,
-                        emailOTPParameters, sendOtpToFederatedEmail);
-            } else if (isUserExistence && !isEmailOTPDisableForUser(username, context,
-                    emailOTPParameters)) {
-                if (log.isDebugEnabled()) {
-                    log.debug("Process the EmailOTP optional flow, but user enable emailOTP as second step ");
+
+                if (authenticatedUser == null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Cannot find the authenticated user.");
+                    }
+                    throw new AuthenticationFailedException
+                            ("Authentication failed!. Cannot find the authenticated user.");
                 }
-                email = getEmailValueForUsername(username, context);
-                if (StringUtils.isEmpty(email)) {
-                    if (request.getParameter(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS) == null) {
-                        redirectToEmailAddressReqPage(response, context, emailOTPParameters, queryParams, username);
+
+                if (isLocalUser) {
+                    if (isEmailOTPDisableForUser(username, context, emailOTPParameters)) {
+                        if (isEmailOTPMandatory) {
+                            if (log.isDebugEnabled()) {
+                                log.debug("Email OTP is disabled for the user " + username + ", since email OTP is " +
```

Comment:
Does this log message correct?

Reply:
if the email OTP is configured as mandatory in the server level, then the user can't disable email OTP. Can be corrected as, 'Since email OTP is mandatory, user cannot disable email OTP'